File: Tests/Tests/STPSourceSEPADebitDetailsTest.m
//
//  STPSourceSEPADebitDetails.m
//  Stripe
//
//  Created by Joey Dong on 6/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@import XCTest;

#import "STPTestUtils.h"

@interface STPSourceSEPADebitDetailsTest : XCTestCase

@end

@implementation STPSourceSEPADebitDetailsTest

#pragma mark - Description Tests

- (void)testDescription {
    STPSourceSEPADebitDetails *sepaDebitDetails = [STPSourceSEPADebitDetails decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"SEPADebitSource"][@"sepa_debit"]];
    XCTAssert(sepaDebitDetails.description);
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:@"SEPADebitSource"][@"sepa_debit"] mutableCopy];
        [response removeObjectForKey:field];

        XCTAssertNil([STPSourceSEPADebitDetails decodedObjectFromAPIResponse:response]);
    }

    XCTAssert([STPSourceSEPADebitDetails decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"SEPADebitSource"][@"sepa_debit"]]);
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:@"SEPADebitSource"][@"sepa_debit"];
    STPSourceSEPADebitDetails *sepaDebitDetails = [STPSourceSEPADebitDetails decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(sepaDebitDetails.bankCode, @"37040044");
    XCTAssertEqualObjects(sepaDebitDetails.country, @"DE");
    XCTAssertEqualObjects(sepaDebitDetails.fingerprint, @"NxdSyRegc9PsMkWy");
    XCTAssertEqualObjects(sepaDebitDetails.last4, @"3001");
    XCTAssertEqualObjects(sepaDebitDetails.mandateReference, @"NXDSYREGC9PSMKWY");
    XCTAssertEqualObjects(sepaDebitDetails.mandateURL, [NSURL URLWithString:@"https://hooks.stripe.com/adapter/sepa_debit/file/src_18HgGjHNCLa1Vra6Y9TIP6tU/src_client_secret_XcBmS94nTg5o0xc9MSliSlDW"]);

    XCTAssertNotEqual(sepaDebitDetails.allResponseFields, response);
    XCTAssertEqualObjects(sepaDebitDetails.allResponseFields, response);
}

@end

File: Stripe/StripeiOSTests/STPSourceSEPADebitDetailsTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPSourceSEPADebitDetails.m
//  Stripe
//
//  Created by Joey Dong on 6/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@testable import StripePayments
import XCTest

class STPSourceSEPADebitDetailsTest: XCTestCase {
    // MARK: - Description Tests

    func testDescription() {
        let sepaDebitDetails = STPSourceSEPADebitDetails.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("SEPADebitSource")["sepa_debit"] as? [AnyHashable: Any])
        XCTAssertNotNil(sepaDebitDetails?.description)
    }

    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let requiredFields: [String]? = []

        for field in requiredFields ?? [] {
            var response = STPTestUtils.jsonNamed("SEPADebitSource")["sepa_debit"] as? [AnyHashable: Any]
            response?.removeValue(forKey: field)

            XCTAssertNil(STPSourceSEPADebitDetails.decodedObject(fromAPIResponse: response))
        }

        XCTAssertNotNil(STPSourceSEPADebitDetails.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("SEPADebitSource")["sepa_debit"] as? [AnyHashable: Any]))
    }

    func testDecodedObjectFromAPIResponseMapping() {
        let response = STPTestUtils.jsonNamed("SEPADebitSource")["sepa_debit"] as? [AnyHashable: Any]
        let sepaDebitDetails = STPSourceSEPADebitDetails.decodedObject(fromAPIResponse: response)

        XCTAssertEqual(sepaDebitDetails?.bankCode, "37040044")
        XCTAssertEqual(sepaDebitDetails?.country, "DE")
        XCTAssertEqual(sepaDebitDetails?.fingerprint, "NxdSyRegc9PsMkWy")
        XCTAssertEqual(sepaDebitDetails?.last4, "3001")
        XCTAssertEqual(sepaDebitDetails?.mandateReference, "NXDSYREGC9PSMKWY")
        XCTAssertEqual(sepaDebitDetails?.mandateURL, URL(string: "https://hooks.stripe.com/adapter/sepa_debit/file/src_18HgGjHNCLa1Vra6Y9TIP6tU/src_client_secret_XcBmS94nTg5o0xc9MSliSlDW"))

        XCTAssertEqual(sepaDebitDetails!.allResponseFields as NSDictionary, response! as NSDictionary)
    }
}
